### PR TITLE
fix: only install @opencode-ai/plugin into config dirs that can import it

### LINKED ISSUE #1293: fix: fresh config startup stalls on unnecessary plugin dependency installs
A fresh configuration can trigger an unnecessary in-process Arborist installation of `@opencode-ai/plugin`, delaying HTTP requests and streamed output for minutes. PR #1292 records the 0.11.0 reproduction: a consent request waited about five minutes, and a first prompt took 149 seconds versus 33 seconds after dependencies were present.

To reproduce, run the compiled CLI with a fresh HOME/XDG environment, PURE unset, and no local plugins or tools. Start `serve` and request an instance endpoint such as `/config` or `/provider`. Bare config directories should not install packages, and instance requests should remain responsive.

The config callsite changed from an external BunProc installer to in-process Npm.Service in [9e4356302f](https://github.com/AltimateAI/altimate-code/commit/9e4356302f148409bcb9e7fca3dc17e96acb2844), June 23, 2026, later merged through #964. The first stable release with that code was 0.9.1; this identifies the code boundary, not a performance bisect of historical binaries. Related upstream reports include [anomalyco/opencode#27971](https://github.com/anomalyco/opencode/issues/27971) and [#44684](https://github.com/anomalyco/opencode/issues/44684).

Acceptance:

- Bare config directories and external-only plugins cause no config dependency installs.
- Local tools/plugins still receive dependencies, including declarations merged later, package roots, and symlink aliases.
- Both Config and TuiConfig skip installs under PURE.
- A non-PURE subprocess regression test exercises real instance HTTP routes and rejects npm requests/install artifacts, with bounded timeouts and cleanup.
- The same check runs against the compiled Linux binary before release publication.

Implementation and regression tests: #1292.

### PR DESCRIPTION
**Pull request by anandgupta42**
### Issue for this PR

Closes #1293

### Type of change

- [x] Bug fix

### What does this PR do?

Fresh config loading installed `@opencode-ai/plugin` in every config directory through in-process Arborist. The original 0.11.0 investigation measured a consent request delayed by 5 minutes 8 seconds and a first prompt taking 149 seconds, versus 33 seconds with dependencies already present.

Install only when a directory has local tool/plugin sources, a declared file plugin that resolves there, or an existing `node_modules`. Evaluate after all config sources merge. Use canonical path containment so package roots, `..name` descendants, and symlink aliases work. Both Config and TuiConfig skip installation under PURE; `.gitignore` bootstrapping remains intact.

The config callsite migrated from a child-process installer on [June 23](https://github.com/AltimateAI/altimate-code/commit/9e4356302f148409bcb9e7fca3dc17e96acb2844), merged through #964, and first shipped stably in 0.9.1. Upstream also reports this failure family: [anomalyco/opencode#27971](https://github.com/anomalyco/opencode/issues/27971), [#44684](https://github.com/anomalyco/opencode/issues/44684). The in-process installer remains unchanged for configurations that need local dependencies.

### How did you verify your code works?

- Eight path/config-merge regressions failed before the first review fixes and passed afterward; the new TUI PURE cases also failed before its guard was added.
- Config/plugin/tool suites after the first fixes: **932 pass, 34 skip, 0 fail**.
- Follow-up config/TUI suites: **366 pass, 3 platform skips, 0 fail**. Serve subprocess tests: **7 pass**.
- Added a non-PURE startup check that loads real config/provider HTTP endpoints, waits for plugin initialization, and rejects npm requests, install artifacts, and background install failures. HTTP reads and server shutdown are bounded.
- Mutation-tested the new check: restoring unconditional installs fails in both source and compiled executions. Rebuilt the fixed macOS arm64 binary and verified the same check passes.
- The startup check runs in the PR subprocess suite and against the compiled Linux binary in push-to-main CI and release builds, before artifact upload/publication.
- Repository typechecking, strict fork-marker checks, workflow YAML parsing, and `git diff --check` pass.

### Screenshots / recordings

Not applicable; config loading and test coverage only.

### Checklist

- [x] I have tested my changes locally
- [x] I have not included unrelated changes in this PR

## Summary by CodeRabbit

- **Bug Fixes**
  - Reduced unnecessary dependency installation for configuration directories without local plugins, tools, or existing dependencies.
  - Dependency installation now recognizes eligible local plugins across all configuration sources.
  - External file plugins and symlinked plugins resolving outside the configuration directory are excluded.
  - Pure mode consistently prevents automatic plugin dependency installation.
  - Improved handling for plugin paths that cannot be resolved on disk.
  - Improved fresh-start reliability by preventing unnecessary registry access and dependency-installation errors.
  - Added stronger cold-start checks for compiled CLI builds.

---

> [!NOTE]
> **Medium Risk**
> Changes core config-load behavior and when npm runs on startup; mitigated by broad unit/integration tests and release binary checks, but edge cases around symlinked or cross-source file plugins could still differ from prior always-install semantics.
> 
> **Overview**
> **Stops unconditional `@opencode-ai/plugin` installs** in every config directory on startup, which had been blocking cold starts for minutes. Installs now run only when a directory has local tool/plugin sources, an existing `node_modules`, or a merged `file://` plugin that resolves inside that dir; **PURE** still skips installs entirely. The decision runs **after** all config sources (inline, account, managed, later dirs) are merged, and **TuiConfig** uses the same gate.
> 
> **CI and regression coverage:** a new **cold-start** job (PRs touching TypeScript + push) and a **release** step run `fresh-start.test.ts` against the **compiled** Linux binary via `OPENCODE_TEST_CLI`, with workflow comments clarifying why that exception is safe versus the general subprocess suite.
> 
> **Test harness:** `opencode serve` subprocess helpers gain deterministic **stop**, full **stderr** capture (EOF/truncation-safe), and SIGKILL fallback on teardown; unit tests cover stderr capture.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3beb7bd3ce6c222be0051fe5a3354476ff9609b9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by anandgupta42** [comment]
Reviewed the original head `67407c1a3a` and pushed fixes in [cef44a6ac4](https://github.com/AltimateAI/altimate-code/commit/cef44a6ac4).

**Standards:** no hard violations found. The recording Npm fixtures now have independent state per test.

**Spec:** found and fixed three functional regressions:

1. **P2 — Plugins from later config sources missed dependency installation.** The gate ran inside the directory loop before inline, account, managed, and later-directory declarations were all known. A local plugin introduced afterward could lose its required `@opencode-ai/plugin` install. Install scheduling now runs once after all plugin configuration merges, preserving the PURE guard and dependency-wait fibers.
2. **P2 — Valid local package paths were rejected.** `rel !== ""` excluded a package plugin rooted at the config directory itself (`plugin: ["."]` with `package.json`); `startsWith("..")` also excluded ordinary descendants such as `..plugins/hello.ts`. The gate now uses core `FSUtil.contains`, which handles equality and separator boundaries correctly.
3. **P2 — Symlink aliases incorrectly skipped dependencies.** A symlink config directory and a physical plugin path, or the reverse, can refer to the same dependency location while failing lexical containment. The gate now compares canonical paths, with lexical fallback for unresolved paths.

**Validation:** all eight new regression cases failed against the original production code while its nine existing gate tests passed; all 17 pass with the fixes. Added six independent TuiConfig gate tests as well.

- Config/plugin/tool suites: **932 passed, 34 skipped, 0 failed**; 16 snapshots.
- Repository typechecking: **13 tasks passed**, including the normal pre-push hook.
- Strict fork-marker check, tracker-reference check, and `git diff --check`: passed.
- Compiled macOS arm64 binary in a fresh isolated HOME, with PURE unset: health ready in **6.432 s**, `/config` returned 200 in **17 ms**, `/provider` in **180 ms**. No dependency-install artifacts appeared. Telemetry, default plugins, and model refresh were disabled for this smoke test.

A second review of the production fixes found no additional blocking issues. The in-process installer remains unchanged, so configurations that require local dependencies can still incur its installation cost.

File: packages/opencode/src/config/plugin.ts [hunk 1]
```diff
@@ -37,6 +40,36 @@ export function pluginOptions(plugin: ConfigPluginV1.Spec): ConfigPluginV1.Optio
...
+  try {
+    if (Glob.scanSync(SOURCE_GLOB, { cwd: dir, dot: true, symlink: true }).length > 0) return true
+  } catch {
+    // An unreadable dir cannot hold importable sources; fall through to the declared specs.
+  }
+  return (plugins ?? []).some((plugin) => {
+    const spec = pluginSpecifier(plugin)
+    if (!spec.startsWith("file://")) return false
+    try {
+      const rel = path.relative(dir, fileURLToPath(spec))
```
[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/src/config/plugin.ts]: Addressed in cef44a6ac4. The gate reuses core FSUtil.contains after canonicalizing paths. I used that helper because it handles equality and ..name descendants correctly; the older Filesystem.contains helper still rejects names beginning with two dots. Regression tests cover package roots, two-dot names, and symlink aliases.

File: packages/opencode/src/config/config.ts [hunk 2]
```diff
@@ -575,7 +575,10 @@ export const layer = Layer.effect(
           // workspace or package cache, so this install fails+retries against the sandbox network and
           // waitForDependencies() (Fiber.join) then HANGS the process on exit — every subprocess test
           // that runs a prompt times out. PURE already means "no external plugin discovery + install".
-          if (!Flag.OPENCODE_PURE) {
+          // Outside PURE, only install where a local plugin/tool source can import the package; the
+          // unconditional in-process arborist reify froze fresh installs for minutes (see
+          // ConfigPlugin.needsDependencies).
+          if (!Flag.OPENCODE_PURE && ConfigPlugin.needsDependencies(dir, result.plugin)) {
```
[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/src/config/config.ts]: Addressed in cef44a6ac4. Installation is scheduled after all config sources merge. Four regression tests cover inline, account, managed, and later-directory declarations, each asserting exactly one install in the required directory.

File: packages/opencode/src/config/tui.ts [hunk 3]
```diff
@@ -273,7 +273,10 @@ export const layer = Layer.effect(
     const data = yield* loadState({ directory, worktree })
     // altimate_change end
     const deps = yield* Effect.forEach(
-      data.dirs,
+      // altimate_change start — upstream_fix: same lazy gate as Config; the unconditional in-process
+      // arborist install froze fresh installs for minutes (see ConfigPlugin.needsDependencies).
+      data.dirs.filter((dir) => ConfigPlugin.needsDependencies(dir, data.config.plugin)),
```
[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/src/config/tui.ts]: Addressed in cef44a6ac4 and extended in 45f63c4f3d. TuiConfig has an independent recording-Npm matrix for bare, npm, file, tool, existing node_modules, and external-file scenarios; every scenario now also runs under PURE. Each test has its own recorder and waits for dependency completion.

[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/src/config/tui.ts]: Fixed in 45f63c4f3d: TuiConfig now applies !Flag.OPENCODE_PURE before the dependency gate. The local-file, tool, and existing-node_modules PURE cases failed before this change and pass afterward. Added matching Config PURE tests plus a non-PURE startup regression in PR CI and compiled release checks.

[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/src/config/tui.ts]: Addressed in cef44a6ac4 and extended in 45f63c4f3d. test/config/tui.test.ts independently tests the gate with recording Npm layers across six source shapes in both ordinary and PURE modes, and waits for dependency completion. Follow-up config/TUI suites: 366 passed, three platform skips.

**Comment by anandgupta42** [comment]
Pushed follow-up [45f63c4f3d](https://github.com/AltimateAI/altimate-code/commit/45f63c4f3d): TuiConfig now honors PURE, and the startup regression has automated coverage in PR CI and compiled main/release checks. Existing review threads have been addressed with fixes or test references. Linked #1293 and updated the PR template.

**When introduced:** the fork's config callsite switched from a child-process BunProc install to in-process Arborist in [9e4356302f](https://github.com/AltimateAI/altimate-code/commit/9e4356302f148409bcb9e7fca3dc17e96acb2844), **June 23, 2026**. It entered main through [#964](https://github.com/AltimateAI/altimate-code/pull/964) on July 8. The first published beta located was **0.9.0-beta.2** (July 3 UTC); the first stable was **0.9.1** (July 9 UTC / July 8 Pacific), per [npm publication metadata](https://registry.npmjs.org/@altimateai%2faltimate-code). This is the code/release boundary; I have not performance-bisected every historical binary. The multi-minute risk predates 0.11.0.

**Upstream is affected too.** Its [Arborist migration #18308](https://github.com/anomalyco/opencode/pull/18308) merged April 1. [#27971](https://github.com/anomalyco/opencode/issues/27971) reports a 298-second request stall; [#31463](https://github.com/anomalyco/opencode/issues/31463) reports cold-cache hangs; [#44684](https://github.com/anomalyco/opencode/issues/44684) reports headless startup hangs. [Timeout PR #47430](https://github.com/anomalyco/opencode/pull/47430), whose author measured 181–370-second stalls, was still unmerged when checked. Warm dependency/cache fast paths explain why many existing installations avoid the expensive path; they do not make upstream immune.

**Why checks missed it:** [3501636dda](https://github.com/AltimateAI/altimate-code/commit/3501636ddab6da5bb01aa5d292a140a053d194b3) already documented the fresh-HOME binary hang on June 26 and bypassed installs under PURE. The subprocess harness still defaulted to PURE, while the explicit release-binary smoke only ran `--version`, which never loads instance config.

**Prevention now added:**

- PURE coverage for both loaders, plus an escaping-symlink negative case.
- Real non-PURE subprocess startup with isolated HOME/XDG directories and a rejecting local npm registry. Requests load config and provider routes. A dependency-free external plugin proves config dependency fibers have completed before assertions, avoiding a race with detached installs.
- Fail on package requests, dependency artifacts, failed background installs, HTTP timeouts, or a plugin that never initializes. Bounded graceful shutdown escalates to SIGKILL.
- Same smoke test against the compiled Linux binary before release artifact upload; source version runs in the regular PR subprocess suite.

Mutation proof: restoring unconditional installs fails the source test on two npm requests and the compiled test on failed background installs. The rebuilt fixed binary passes. Follow-up validation: **366 config/TUI tests + 7 serve tests passed**, **3 platform skips**; compiled smoke passed; all **13 typecheck tasks** passed. Marker/tracker checks and workflow YAML validation passed.

Full cited investigation is saved in the review worktree and the altimate Obsidian vault. It also records a separate CLI/plugin version mismatch for local-plugin users; that is a compatibility concern, not an established cause of the measured stall, and this PR does not change package-version selection.

**Comment by anandgupta42** [comment]
Remaining review items addressed in the latest commit:

- **PURE predicate duplicated** (Kilo, `tui.ts`/`config.ts`): both call sites now use `ConfigPlugin.shouldInstallDependencies(dir, plugins)`, which folds in the `!Flag.OPENCODE_PURE` check with a docstring on why PURE skips.
- **Cold-start regression only ran on push** (cubic, `ci.yml`): moved into its own `cold-start-regression` job gated like `typescript` (PRs touching TS, plus every push), so a PR reintroducing the stall fails PR CI. The `OPENCODE_TEST_CLI` NOTE now cross-references this single bounded exception instead of contradicting it (Kilo).
- **stderr tail buffer unbounded / `stderr()` races the drain fiber** (Kilo, cubic, `cli-process.ts`): bounded at 64 KB, and `stderr()` now returns an Effect that waits for the drain to go quiet (20 ms quiet interval, 500 ms cap), so the `not.toContain` assertion in `fresh-start.test.ts` is deterministic.
- **Duplicated hardcoded binary path** (cubic, `release.yml`): the smoke-test step exports its `find`-resolved path and the cold-start step consumes it.

Verification: `bun test test/config test/cli/serve` 269 pass, `fresh-start.test.ts` under `--timeout 90000` passes, typecheck clean, marker guard ok, both workflow files parse.

**Comment by anandgupta42** [comment]
@cursor review

**Comment by anandgupta42** [comment]
Round 3 (CodeRabbit `cli-process.ts:108`, cubic `cli-process.ts:110`): the sleep-polling `settledStderr` is gone. `ServeHandle` now has `stop()` (SIGTERM, SIGKILL fallback, idempotent with the scope finalizer) and `stderr()` awaits process exit and `Fiber.join`s the drain fiber before reading, so the read is complete by construction rather than by a quiet interval or a chunk count that the 64 KB trim can hold steady. `fresh-start.test.ts` calls `stop()` before the `not.toContain` assertion. `bun test test/cli/serve --timeout 90000`: 7 pass.

**Comment by anandgupta42** [comment]
cubic `cli-process.ts:196`: the `stderr()` doc now says what it returns, the retained tail (last `STDERR_TAIL_BYTES`), and notes that a `not.toContain` check is only as strong as that window.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -639,6 +639,14 @@ jobs:
           ALTIMATE_BASE_GATEWAY_URL: https://gateway.test
           MODELS_DEV_API_JSON: test/tool/fixtures/models-api.json
 
+      # altimate_change start — --version and PURE-mode tests never exercise ordinary config installs.
+      - name: Cold-start config regression (compiled, non-PURE)
+        working-directory: packages/opencode
+        env:
+          OPENCODE_TEST_CLI: ${{ github.workspace }}/packages/opencode/dist/@altimateai/altimate-code-linux-x64/bin/altimate-code
```
[the remark this answers is not in the document]
Comment (anandgupta42) [reply .github/workflows/ci.yml]: Addressed in f88cc2c155: the general subprocess workflow NOTE now points to the dedicated compiled cold-start exception. The broader run/mock suite retains its existing source execution policy.

File: packages/opencode/src/config/tui.ts [hunk 5]
```diff
@@ -273,7 +273,10 @@ export const layer = Layer.effect(
     const data = yield* loadState({ directory, worktree })
     // altimate_change end
     const deps = yield* Effect.forEach(
-      data.dirs,
+      // altimate_change start — upstream_fix: same lazy gate as Config; the unconditional in-process
+      // arborist install froze fresh installs for minutes (see ConfigPlugin.needsDependencies).
+      data.dirs.filter((dir) => !Flag.OPENCODE_PURE && ConfigPlugin.needsDependencies(dir, data.config.plugin)),
```
[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/src/config/tui.ts]: Addressed in f88cc2c155: Config and TuiConfig now call ConfigPlugin.shouldInstallDependencies, which centralizes the PURE skip and local dependency predicate. Existing PURE/config/TUI regressions cover both callers.

File: packages/opencode/test/lib/cli-process.ts [hunk 6]
```diff
@@ -424,6 +433,8 @@ export function withCliFixture<A, E>(
           proc.kill()
         },
         exited: proc.exited as Promise<number>,
+        // altimate_change — expose the already-drained diagnostic output.
+        stderr: () => stderrChunks.join(""),
```
[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/test/lib/cli-process.ts]: Addressed in 3beb7bd3ce: stderr capture retains at most 64K UTF-16 characters for diagnostics. Complete-output assertions now fail if any output was truncated, so an earlier installation error cannot disappear into a clean retained tail. Regression coverage explicitly exercises that case.

File: packages/opencode/test/lib/cli-process.ts [hunk 7]
```diff
@@ -72,12 +77,47 @@ function forkStderrDrain(stream: ReadableStream<Uint8Array>, into: string[]) {
...
 
+// `forkStderrDrain` appends off its own I/O callback, which can still be catching up with data the
+// child already wrote (and that a caller's own await — e.g. an HTTP round trip to the same process —
+// already observed the effect of) by the time a test wants to assert on stderr content. Poll until
+// the buffer goes quiet for one full interval, bounded so a genuinely idle pipe can't hang the test;
+// this makes `not.toContain(...)` assertions deterministic instead of racing the drain fiber.
+function settledStderr(chunks: string[]) {
+  return Effect.gen(function* () {
+    let previous = -1
+    while (chunks.length !== previous) {
```
[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/test/lib/cli-process.ts]: Addressed by the EOF synchronization in 2e54ba41af and hardened in 3beb7bd3ce. The reader no longer polls chunk count or inactivity: stop() awaits process exit and stderr() joins the drain through EOF. Truncation and read errors fail the assertion. The controlled-stream regression failed against the old quiet-period implementation and passes with EOF synchronization.

File: packages/opencode/test/lib/cli-process.ts [hunk 8]
```diff
@@ -157,6 +188,13 @@ export type ServeHandle = {
   readonly kill: () => void
   // Resolves with the exit code once the process exits. Bun returns a number.
   readonly exited: Promise<number>
+  // altimate_change — let startup checks detect failed background work as well as HTTP failures.
+  // Terminates the child (SIGTERM, SIGKILL after 5 s) and resolves with its exit code. Idempotent;
+  // the scope finalizer performs the same shutdown if a test never calls it.
+  readonly stop: () => Effect.Effect<number>
+  // Complete stderr, available deterministically only after the child exited: awaits exit and joins
+  // the drain fiber (see drainedStderr). Call stop() first in a test that wants to assert on it.
```
[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/test/lib/cli-process.ts]: Fixed behavior as well as the contract in 3beb7bd3ce. The bounded tail remains available for timeout diagnostics, but stderr() rejects truncated capture instead of returning partial output to a negative assertion. The regression writes an installation error, then enough clean output to evict it, and verifies complete capture fails.

File: packages/opencode/test/lib/cli-process.ts [hunk 9]
```diff
@@ -72,12 +77,38 @@ function forkStderrDrain(stream: ReadableStream<Uint8Array>, into: string[]) {
...
 }
 
+// `forkStderrDrain` appends off its own I/O callback, so reading the buffer while the child is
+// alive races the drain fiber (test/AGENTS.md: no sleep-based waiting for a forked fiber). The only
+// deterministic read is after the child has exited: the pipe closes, the drain stream ends, and
+// joining the fiber proves every chunk landed. `ServeHandle.stop()` provides that sequence.
+function drainedStderr(exited: Promise<number>, drain: Fiber.Fiber<void>, chunks: string[]) {
+  return Effect.gen(function* () {
+    yield* Effect.promise(() => exited)
+    yield* Fiber.join(drain)
```
[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/test/lib/cli-process.ts]: Addressed in 3beb7bd3ce: the replacement capture helper uses Effect.fn("CliStderr.capture"), and stop uses Effect.fn("opencode.serve.stop"). All 13 typecheck tasks and the 10 serve/stderr tests pass after this adjustment.

**Comment by anandgupta42** [comment]
Pushed `3beb7bd3ce` to close the remaining startup-test false-negative paths, building on the concurrent shutdown fix in `2e54ba41af`.

The test now waits for process exit and stderr EOF. Bounded capture fails if output was truncated or unreadable, so a missing error in a partial tail cannot make the regression pass. The new controlled-stream tests cover delayed output after the former quiet interval, an early error evicted by the cap, and pipe read failure. This commit changes test infrastructure only.

Validation:

- The quiet-interval regression failed against the old reader and passes with EOF synchronization.
- Serve, ACP, and stderr suite: **25 passed, 0 failed**. After the final rebase/Effect naming adjustment, all **10 serve/stderr tests passed** again.
- All **13 typecheck tasks passed**, including the final pre-push check.
- Native macOS arm64 build and source/compiled non-PURE cold-start E2Es passed.
- Mutation proof: restoring unconditional installation made the source E2E fail on registry requests and the compiled E2E fail on the background-install error. Restoring the gate and rebuilding made both pass.
- Marker, tracker-leak, and whitespace checks passed. Independent Standards and Spec reviews found no remaining blockers after the Effect naming adjustment.

The fix has been pushed; merge still depends on the new-head CI results and required reviewer approval.
